### ISSUE TITLE
Bug in FTPFactory is avoiding the change of anonymous username

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3390]
|<img alt="boltrix's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| boltrix reported|
|-|-|
|Trac ID|trac#3390|
|Type|defect|
|Created|2008-08-14 08:37:48Z|

There is a bug in the current revision r24576, which avoid user from setting a different anonymous username. Here is a link to the buggy code: http://twistedmatrix.com/trac/browser/tags/releases/twisted-8.1.0/twisted/protocols/ftp.py#L1247

I have attached a patch for that bug, and another patch with a test suite for checking that the bug is gone.

Attachments:

* [ftp_anonymous_username.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f79/f79cabe72086fbd294ca4025842d1597aaf27094/c2d5326915e3d2867361819749724f9ac1d70533.patch) (462 bytes) - added by boltrix on 2008-08-14 08:38:58Z - Patch for twisted.protocol.ftp
* [ftp_anonymous_username_test_case.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f79/f79cabe72086fbd294ca4025842d1597aaf27094/03767bb852a8db449e544eaea3114539661a0534.patch) (1947 bytes) - added by boltrix on 2008-08-14 08:39:42Z - Patch for twisted.test.test_ftp test case, adding the test for different anonymous username
* [ftp_anonymous_username_test_case.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f79/f79cabe72086fbd294ca4025842d1597aaf27094/d6e3300a116bbe3036c2294f61266cea774b1064.patch) (2190 bytes) - added by boltrix on 2008-08-14 08:58:50Z - A patch to the twisted.test.test_ftp allowing it to test other usernames for anonymous login

<details><summary>Searchable metadata</summary>

```
trac-id__3390 3390
type__defect defect
reporter__boltrix boltrix
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1218703068000000 1218703068000000
changetime__1218965403000000 1218965403000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="boltrix's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|boltrix commented|
|-|-|

Just adding my e-mail, just in case someone have it fixed and want to get in touch with me: py (dot) adriano <at> gmail [dot] com

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to boltrix|
|-|-|

Thanks for the patches, this is almost perfect. I have only the following comments:
 * the clientFactory attribute of `FTPAnonymousTestCase` is not necessary, it will be inherited by the parent
 * instead of passing an attribute to setUp, I would prefer that you use a class attribute userAnonymous = 'anonymous' in `FTPServerTestCase`, overriden in `FTPAnonymousTestCase`
 * testAnonymousLogin should be named test_anonymousLogin, and needs a docstring.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="boltrix's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|boltrix commented|
|-|-|

Thanks a lot for the comments!
I did write the fixes commented, and I'm attaching a new patch for it now. Let me know if I'm still missing something!

**Comment by twisted-trac** [comment]
|<img alt="boltrix's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|boltrix removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set status to `closed`|
|-|-|

(In [24609]) Make FTPFactory uses its userAnonymous parameter instead of ignoring it.

Author: boltrix
Reviewer: therve
Fixes [#3390](https://github.com/twisted/twisted/issues/3390)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
